### ISSUE TITLE
Discussion: about Roll Number

### ISSUE THREAD

**Issue by dzhuang** [inducer/relate#99]
Before starting a new course, it is often the case that we get the name list of students with their unique roll number (or register number, or student ID) from the official teaching system of the college, while email of students are not known before hand.  I know many college use official email address like ``@illinois.edu`` as roll number, but there are still many use other format of roll number.

If we add a roll_number field in the user model, it may bring many convenience for the relate system:

 1. Preapproving student's enrollment before we can collected their email address.
 1. Knowing the real name behind a username/email, which consequently ease the impersonation of students, which is especially helpful when some profiles are not filled in, or having typo in profiles.
 1. Finding out which students did not enroll the Relate course as required.
 1. Enabling roll call in Relate system even some students did not enrolled the Relate course.
 1. Helpful when export/import of students grades with roll number as primary key.

**Comment by inducer** [comment]
I think this is a reasonable idea. Practically, however, there are a few wrinkles. First of all, relate would need to migrate to a custom user model within Django. A patch for this exists:

https://github.com/inducer/relate/tree/custom-user-model

Or, alternatively, we could bolt this field into the user status model. But even if this wrinkle is resolved, it can be somewhat tricky to make sure that a student who is claiming to be associated with one student record actually is who he is claiming to be. Ownership of an email is easy to prove, but ownership of a student ID much less so. Ideally, relate should integrate with an existing single-sign-on system at the University, but no canned facility for that exists as yet. But aside from these difficulties, having the field and using it in the situations you mention should be easy.

**Comment by dzhuang** [comment]
Yes, I know about the migration wrinkle, and actually I've temporarily added the field into user status on my Relate instance, along with restriction that once the unique student ID are submitted, it is not changeable. 

I must admit that official email is much better than student ID in confirming it is actually the student who is accessing the course. However for many other universities, at least for the case in China, do not provide such an official mail service to student. What's more, I think if we bind email to student id, with the option that the course must be enroll by specific email suffix, we are still using email to justify that the student is actually used by that ID. Am I right?

**Comment by inducer** [comment]
@lukeolson is running into a similar issue where he needs another bit of credible identity information besides the email address. Student ID would be ideal, First/Last would also be good. So we need a way of getting this information into RELATE somehow.

**Comment by dzhuang** [comment]
I'll try to create a PR if you don't mind, as the application of Student ID on my instance is buggy and the code is messy :-(

**Comment by inducer** [comment]
I am wondering if we should just bite the bullet and finally go to the custom user model. It'll be a bit painful, but I imagine we'll be happier after. We could also get rid of the `UserStatus` at this point.

In terms of designing this, one way I could imagine is to have a flag in the user database that indicates whether their student ID and their first/last name are "verified". As soon as they are verified, we take away the students' ability to edit them. And they would get verified through some CSV import process, possibly the same as the enrollment preapprovals. (so this would have to be changed to actually create user accounts)

**Comment by dzhuang** [comment]
I think its a great idea to remove user status model and suffer the pain of creating and migrating to a customized user model. If that is done, I imagine there're some more funtionality needed, for example, forgetting username, forgetting which email is used in registration, complaint about abusive registration of student ID.

**Comment by inducer** [comment]
bd69504642ab0865c0c1b23351a2678d14087e56 makes some progress on this. Notably, the custom user model is now merged into master. And this introduces an 'Institutional ID' along with flags indicating whether they're verified.

**Comment by dzhuang** [comment]
Sorry for late response, as gmail is difficult to access in some country and I didn't get the notification email until just now. Thank you for the great improvements !

**Comment by inducer** [comment]
We're done here, right?
